test/c/test_sparse_observable.c

```diff
@@ -0,0 +1,650 @@
+ */
+int test_num_terms() {
+    int result = Ok;
+    uint64_t num_terms;
+
+    QkSparseObservable *zero = qk_obs_zero(100);
+    num_terms = qk_obs_num_terms(zero);
+    if (num_terms != 0) {
+        result = EqualityError;
+    }
+    qk_obs_free(zero);
+
+    QkSparseObservable *iden = qk_obs_identity(100);
+    num_terms = qk_obs_num_terms(iden);
+    if (num_terms != 1) {
+        result = EqualityError;
+    }
+    qk_obs_free(iden);
+
+    return result;
+}
+
+/**
+ * Test getting the number of qubits in an observable.
+ */
+int test_num_qubits() {
+    int result = Ok;
+    uint32_t num_qubits;
+
+    QkSparseObservable *obs = qk_obs_zero(1);
+    num_qubits = qk_obs_num_qubits(obs);
+    if (num_qubits != 1) {
+        result = EqualityError;
+    }
+    qk_obs_free(obs);
+
+    QkSparseObservable *obs100 = qk_obs_zero(100);
+    num_qubits = qk_obs_num_qubits(obs100);
+    if (num_qubits != 100) {
+        result = EqualityError;
+    }
+    qk_obs_free(obs100);
+
+    return result;
+}
+
+/**
+ * Test adding an individual term to an observable.
+ */
+int test_custom_build() {
+    u_int32_t num_qubits = 100;
+    QkSparseObservable *obs = qk_obs_zero(num_qubits);
+
+    complex double coeff = 1;
+    QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+    uint32_t indices[3] = {0, 1, 2};
+    QkSparseTerm term = {coeff, 3, bit_terms, indices, num_qubits};
+
+    qk_obs_add_term(obs, &term);
+    qk_obs_add_term(obs, &term);
+
+    double tol = 1e-6;
+    QkSparseObservable *simplified = qk_obs_canonicalize(obs, tol);
+
+    uint64_t num_terms = qk_obs_num_terms(obs);
+    uint64_t num_terms_simplified = qk_obs_num_terms(simplified);
+
+    qk_obs_free(obs);
+    qk_obs_free(simplified);
+
+    if (num_terms != 2 || num_terms_simplified != 1) {
+        return EqualityError;
+    }
+    return Ok;
+}
+
+/**
+ * Test getting the terms in an observable.
+ */
+int test_term() {
+    uint32_t num_qubits = 100;
+    QkSparseObservable *obs = qk_obs_identity(num_qubits);
+
+    QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+    uint32_t qubits[3] = {0, 1, 2};
+    complex double coeff = 1 + I;
+
+    QkSparseTerm term = {coeff, 3, bit_terms, qubits, num_qubits};
+    int err = qk_obs_add_term(obs, &term);
+
+    if (err != 0) {
+        return RuntimeError;
+    }
+
+    // some placeholders to store the results
+    int nnis[2] = {-1, -1};
+    int bits[3] = {-1, -1, -1};
+    int indices[3] = {-1, -1, -1};
+
+    uint64_t num_terms = qk_obs_num_terms(obs);
+    for (uint64_t i = 0; i < num_terms; i++) {
+        QkSparseTerm view;
+        qk_obs_term(obs, i, &view);
+        size_t nni = view.len;
+        nnis[i] = nni; // store to compare later
+
+        for (uint32_t n = 0; n < nni; n++) {
+            // this loop is only called once, so we can use ``n`` to index here
+            bits[n] = view.bit_terms[n];
+            indices[n] = view.indices[n];
+        }
+    }
```

Comment:
There's some pretty weird typing going on here.  Everything's `int` up top, then it's assigned various other values, etc.  Might be good to get some linting set up on this.

Reply:
Fixed in 624c3dd